adds scoped values credential to support multiple endpoints.

ghyatzo opened the pull request:
Hello,

I had wrote my script for downloading CDS data, before checking if it existed already.
Funnily enough our implementations are eerily similar.
There are some slight differences that I would like to propose as improvements/additions to this package so we can have a single point.

This is the first of 3 pull requests, one for each proposal so they can be addressed independently.

Ok with that out of the way:
I would like to propose the use of scoped values for the API credentials. As well as a bit better credential retrieval mechanism that also checks for environmental variables.

This allows to use different credentials for each request, for example:
```julia
cred1 = CDScredentials("path/to/first/set/of/credentials")
cred2 = CDScredentials("path/to/second/set/of/credentials")

with( auth => cred1) do
    CDSAPI.retrieve(name, request, target)
end

with( auth => cred2) do
    # makes the requests with a different token
    CDSAPI.retrieve(name, request, target)
end

# If none are provided, we look for credentials in the environmental variables
# and then in the default "~/.cdsapirc"
CDSAPI.retrieve(name, request, target)
```

ghyatzo commented on src/CDSAPI.jl:
```diff
@@ -2,6 +2,50 @@ module CDSAPI
+    if any(isempty, (url, token))
+        dotrc = joinpath(homedir(), ".cdsapirc")
+        if !isfile(dotrc)
+            error("""
```
I mulled this over as well, the only two candidates I thought of were `ArgumentError` or `SystemError`... but either don't sound very convincing to be honest.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -2,6 +2,50 @@ module CDSAPI
+const auth = ScopedValue(Dict("url" => "", "key" => ""))
+
+"""
+    CDScredentials()
```
```suggestion
    credentials()
```

Please find/replace all occurrences of the old name.

ghyatzo commented on src/CDSAPI.jl:
```diff
@@ -14,13 +58,12 @@ The client periodically requests the status of the retrieve request.
-            key, val = strip.(split(line, ':', limit=2))
-            creds[key] = val
-        end
+
+    if any(isempty, values(auth[]))
+        # grabs default credentials if none are provided.
+        auth[] = CDScredentials()
     end
+    creds = auth[]
```
Yeah, probably putting it into the credential helper directly makes more sense.
But some sort of check is needed anyway, since we need to differentiate when we call with `with(auth => cred)` and when we call normally.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -2,6 +2,50 @@ module CDSAPI
+"""
+    CDScredentials()
+
+Checks the default location for the cds credentials
```
```suggestion
Attempt to find CDS credentials using different methods.
```

Please document these methods, particularly their precedence order.

ghyatzo commented on src/CDSAPI.jl:
```diff
@@ -2,6 +2,50 @@ module CDSAPI
 
 using HTTP
 using JSON
+using Base.ScopedValues
```
ScopedValues are in Base since 1.11, the ScopedValues.jl is compatible for Julia 1.8+.
I have to admit, I am not sure how to include `ScopedValues.jl` in the project but only for julia versions between 1.11 and 1.8...

juliohm replied:
I commented above on how to install ScopedValues.jl

juliohm commented on Project.toml:
```diff
@@ -10,7 +10,7 @@ JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
 HTTP = "1"
 JSON = "0.21"
-julia = "1"
+julia = "1.11"
```
```suggestion
julia = "1.10"
```
We need to support LTS. You can activate the folder with `Pkg.activate(".")` and then `Pkg.add("ScopedValues")`. That way you can `using ScopedValues` instead of the one in Base.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -2,6 +2,63 @@ module CDSAPI
+const auth = ScopedValue(Dict("url" => "", "key" => ""))
+
+"""
+    credentials()
+
+Attempt to find CDS credentials using different methods.
+The order of precedence is:
+    - direct credentials provided via a specific file.
+    - environmental variables: `CDSAPI_URL` and `CDSAPI_KEY`
+    - default credential file location: `$HOME/.cdsapirc`
+
+The expected file format is:
+url: https://yourendpoint
+key: your-personal-api-token
```
```suggestion
    credentials()

Attempt to find CDS credentials using different methods:

    1. direct credentials provided via a specific file
    2. environmental variables `CDSAPI_URL` and `CDSAPI_KEY`
    3. credential file in home directory `$HOME/.cdsapirc`

A credential file is a text file with two lines:

url: https://yourendpoint
key: your-personal-api-token
```

ghyatzo commented:
I'll add some examples of this new usage in the readme later

juliohm commented on src/CDSAPI.jl:
```diff
@@ -3,6 +3,72 @@ module CDSAPI
 using HTTP
 using JSON
 
+if VERSION >= v"1.11"
+    using Base.ScopedValues
+elseif VERSION >= v"1.8"
+    import Pkg
+    Pkg.activate(".")
+    Pkg.add("ScopedValues")
+    using ScopedValues
+end
```
The idea is to run these commands in the source code to modify the Project.toml of the package. There is no problem in adding ScopedValues.jl as a dependency.

ghyatzo replied:
Ok, so then just remove the `Pkg.add("ScopedValues")`?

ghyatzo commented:
OK, I've rebased onto master to have the other PRs.
I'll add some tests and examples in the README

ghyatzo commented:
tests pass locally

juliohm commented:
@ghyatzo the tests are failing. Should we continue the work on this PR?

ghyatzo commented:
Hi, we encountered this issue in a previous PR https://github.com/JuliaClimate/CDSAPI.jl/pull/60#issuecomment-2595223298

Have you tried locally? apparently the PR branches do not have access to the secrets in the CI so the fail I saw were there, but locally with my credentials this branch pass all tests (at least, last time I checked, don't know if something else changed in the mean time)

juliohm commented:
@ghyatzo there are pending review comments. I can try to run things locally again after they are considered.

ghyatzo commented:
I believe I've addressed all the comments. The change requested points back to  https://github.com/JuliaClimate/CDSAPI.jl/pull/58#pullrequestreview-2552217590, otherwise all other conversations have been resolved.
I'll request anothre review

juliohm commented on Project.toml:
```diff
@@ -6,11 +6,13 @@ version = "2.0.3"
 JSON = "0.21"
-julia = "1"
+ScopedValues = "1.3.0"
+julia = "1.8"
```
@ghyatzo this is pending

juliohm commented:
Sorry, I had a review that was pending, that is why it was not showing to you. They comments should be available now.

ghyatzo commented on src/CDSAPI.jl:
```diff
@@ -3,6 +3,71 @@ module CDSAPI
+    using ScopedValues
+end
+
+export with
```
The idea behind the re-export was that the `with` was somewhat part of the API since a user would call the `with` directly like:
```julia
with( CDSAPI.auth => cred1 ) do
    CDSAPI.retrieve(dataset, request, "download.nc")
end
```
without needing to know that `with` comes from the `ScopedValues` pacakge and instead do:
```julia
ScopedValues.with( CDSAPI.auth => cred1 ) do
    CDSAPI.retrieve(dataset, request, "download.nc")
end
```

ghyatzo commented on src/CDSAPI.jl:
```diff
@@ -3,6 +3,71 @@ module CDSAPI
+            """)
+        end
+
+        return credentials(dotrc)
```
You mean not having a separate `credentials()` function?

juliohm replied:
Yes, I believe the encapsulation into a function is not necessary. The outer function is short enough.

ghyatzo replied:
The `credentials(file)` function was meant as a provided parser method for user that wants to quickly load different `.cdsapirc` files, like:

the encapsulation is so that the no argument version of `credentials` checks the various defaults paths one of which is loading the default `.cdsapirc` file.

But an advanced user that already knows which files to load can skip ahead and simply load the file directly using that function.

juliohm replied:
In that case the function needs a distinct name, e.g. `CDSAPI.credentialsfromfile(fname)` and a dedicated docstring. We should probably avoid mentioning it in the README, and avoid making it part of the official API. It can be an internal utility function that advanced users can use at their own risk.

ghyatzo replied:
Maybe change the zero argument method with `CDSAPI.defaultcredentials()` as well then?

juliohm replied:
I believe the name `CDSAPI.credentials()` is fine for a function that attempts different methods to get the credentials. We just need to rename the other function, which does a very specific action loading credentials from a file on disk.

juliohm commented on README.md:
```diff
@@ -16,13 +16,13 @@ Please install the package with Julia's package manager:
+## Basic usage
 
-Make sure your `~/.cdsapirc` file exists. Instructions on how to create the file for your user account can be found
+Make sure your `~/.cdsapirc` file exists or the env vars `CDSAPI_URL` and `CDSAPI_KEY` are set.
```
This paragraph could be expanded to discuss the precedence order of the different credential methods. I believe we could copy/paste the docstring of `CDSAPI.credentials` here with minor adjustments to improve the flow.

ghyatzo commented on test/runtests.jl:
```diff
@@ -85,4 +85,26 @@ datadir = joinpath(@__DIR__, "data")
+            }"""
+
+        creds = CDSAPI.credentials() # grab the default ones
+        with( CDSAPI.auth => creds ) do
```
I'd expect the most practical way to use this feature is storing various pairs into different `.cdsapirc` files. Keeping sets of keys separate seems like a niche scenario within the already niche scenario.

ghyatzo replied:
There can be different API tiers and providers IIRC, which point to different access points.

juliohm replied:
Still, organization-wise, it is strange to set the environment variable as a pair of values. Most users will just want to touch the `CDSAPI.key` using ScopedValues when they adopt this method. If a user relies on credential files, they will not touch these environment variables.

In other words, it is natural to think of credential files as pairs of values as they are defined externally by the official CDS API, but environment variables are an alternative method in this package that can be more fine grained and readable:

```julia
with(CDSAPI.key => "....") do
  ...
end
```

ghyatzo replied:
I mean, as it is right now is (1) that breaks the pattern of not having the two elements of the key joint together, by "virtue" of the format of the env vars.
The issue you raise of possible future changes applies to one parametrization as much as the other, as I pointed out in my comment.
I still think that splitting the two would add unnecessary boilerplate or typing, with little benefit.
But let's not waste more time on this bikeshed :)

juliohm replied:
> But let's not waste more time on this bikeshed :)

Fully agree :) Please adjust the PR with the suggestions and I will pull the changes to run local tests.

ghyatzo replied:
done! I've also fixed the order of getting the credentials, so now every value can be overwritten properly and independently from the other. In order:
* default cred file
* env vars
* scoped vals

juliohm commented on README.md:
````diff
@@ -16,13 +16,24 @@ Please install the package with Julia's package manager:
+    2. environmental variables `CDSAPI_URL` and `CDSAPI_KEY`
+    3. default credential file in home directory `~/.cdsapirc`
+
+A valid credential file is a text file with two lines:
+```
+url: https://yourendpoint
+key: your-personal-api-token
+```
````
I believe we don't need to promote the `credentialsfromfile` approach in the README, it is an internal detail. As suggested in previous comments, we have 3 user-facing methods:

1. scoped values
2. env variables
3. default dotrc file

Anything in the README that deviates from these methods is a source of confusion.

ghyatzo replied:
The specification of the format of the credentials file, was not much to promote the `credentialsfromfile` internal api, but to give an example of what the default credentials file should look like instead. I think it is valuable to give users a reference point of what it should look like to help them better navigare the setup process if they are new to the CDSAPI.

juliohm commented on README.md:
```diff
@@ -82,6 +92,30 @@ Dict{String,Any} with 6 entries:
+dataset = "reanalysis-era5-single-levels"
+request = """ #= some request =# """
+
+_url1, _key1 = CDSAPI.credentialsfromfile("path/to/credential/file1")
+_url2, _key2 = CDSAPI.credentialsfromfile("path/to/credential/file2")
```
Please adjust this example to avoid the `credentialsfromfile` internal function. Sorry for insisting on this point, but we need to be super picky with *public APIs*.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -3,6 +3,81 @@ module CDSAPI
+using ScopedValues
+
+
+const key = ScopedValue("")
+const url = ScopedValue("")
```
Let's rename these constant variables to `KEY` and `URL` so that they are seen as constants in the codebase and so that they feel more consistent with the env variables, which are also upper case.

This change will also eliminate the `_key` and `_url` naming choice in internal implementations. You can just use `key` and `url` without conflict.

ghyatzo commented on README.md:
```diff
@@ -16,13 +16,24 @@ Please install the package with Julia's package manager:
+`CDSAPI.jl` will attempt to use CDS credentials using three different methods with the following priority:
 
-Make sure your `~/.cdsapirc` file exists. Instructions on how to create the file for your user account can be found
+    1. direct credentials provided via a specific file
```
yes. bad wording on my part. My mental model is still constructed around custom files, so it leaked into the readme by mistake.

juliohm commented on test/runtests.jl:
```diff
@@ -85,4 +85,26 @@ datadir = joinpath(@__DIR__, "data")
+            }"""
+
+        creds = CDSAPI.credentials() # grab the default ones
+        with( CDSAPI.auth => creds ) do
```
There is no `CDSAPI.auth` in the proposed methods.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -19,18 +93,13 @@ retrieve(name, params::AbstractString, filename; wait=1.0) =
-        end
-    end
+    
+    _url, _key = credentials()
```
Please remove the underscore of these variable names.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -3,6 +3,80 @@ module CDSAPI
+using ScopedValues
+
+const KEY = ScopedValue("")
+const URL = ScopedValue("")
```
I believe the URL should point to the current CDS API URL by default.

juliohm commented on README.md:
````diff
@@ -82,6 +92,33 @@ Dict{String,Any} with 6 entries:
   "content_length"     => 193660
   "state"              => "completed"
 ```
+# Multiple credentials
````
```suggestion
# Multiple credentials

```

juliohm commented on README.md:
````diff
@@ -16,13 +16,24 @@ Please install the package with Julia's package manager:
 ```
 
-## Usage
+## Basic usage
````
```suggestion
## Basic usage

```

juliohm commented on README.md:
```diff
@@ -16,13 +16,25 @@ Please install the package with Julia's package manager:
-Make sure your `~/.cdsapirc` file exists. Instructions on how to create the file for your user account can be found
+`CDSAPI.jl` will attempt to use CDS credentials using three different methods with the following priority:
+
+    1. direct credentials provided through the scoped values `KEY` and `URL`
```
```suggestion
    1. direct credentials provided through the scoped values `CDSAPI.KEY` and `CDSAPI.URL`
```

juliohm commented on src/CDSAPI.jl:
```diff
@@ -4,6 +4,80 @@ using HTTP
 
+using ScopedValues
+
+const KEY = ScopedValue("")
+const URL = ScopedValue("https://cds.climate.copernicus.eu/api")
```
```suggestion
const URL = ScopedValue("https://cds.climate.copernicus.eu/api")
const KEY = ScopedValue("")
```

ghyatzo replied:
I just noticed, If we provide the default URL in the scoped values, it will always overwrite any other url provided in any other way. I think this is not desirable as it would lead to unexpected behaviour. overwriting through the scoped values should always be explicit in the code.
I propose we reverse this change.

juliohm replied:
Good point. Please feel free to reverse the change 👍🏽

juliohm commented on src/CDSAPI.jl:
```diff
@@ -4,6 +4,80 @@ using HTTP
+
+Attempt to find CDS credentials using different methods:
+
+    1. direct credentials provided via a specific file
+    2. environmental variables `CDSAPI_URL` and `CDSAPI_KEY`
+    3. credential file in home directory `~/.cdsapirc`
```
This docstring is outdated.

juliohm commented on src/CDSAPI.jl:
```diff
@@ -4,6 +4,80 @@ using HTTP
+    key = get(ENV, "CDSAPI_KEY", key)
+
+    # overwrite with ScopedValues provided by user
+
```
```suggestion
```

juliohm commented on src/CDSAPI.jl:
```diff
@@ -4,6 +4,80 @@ using HTTP
+"""
+    credentials(file)
+
+Parse the cds credentials from a provided file
```
```suggestion
Parse the CDS credentials from a provided `file`.
```

juliohm commented on src/CDSAPI.jl:
```diff
@@ -4,6 +4,80 @@ using HTTP
+        end
+    end
+
+    if !(haskey(creds, "url") || haskey(creds, "key")) # we can allow files with only one of the keys.
```
Another undocumented behavior where files follow a different format than the format specified in the README. I would just enforce the presence of both URL and KEY.

ghyatzo commented on src/CDSAPI.jl:
```diff
@@ -4,6 +4,80 @@ using HTTP
+        url = key = ""
+    end
+
+    # overwrite with environmental variables
```
Two methods of control of the credentials are indeed host machine dependent.
And there is not much we can do about it. 
A user would expect that running a request from a machine with a certain combination of env variable would use said env variables.

I am not sure what unpredictable behavior can arise from this. Could you expand on this? Are you suggesting that the env variable would overwrite the file values only if both are present? That would go against the principle of least surprise in my opinion. 

if an environmental variable is set it is supposed to overwrite a specific setting and exactly which one is stated in the documentation. I would find much more unintuitive the fact that I can't overwrite a single value now that the scoped values were changed to act on the two value separately.

juliohm replied:
Fair enough. From the pool of unpredictable behaviors, it seems that the one implemented is more reasonable. It follows the documented behavior on a per-variable basis. So one could use a scoped value for KEY and a environment variable for URL, etc.

ghyatzo commented:
see: https://github.com/JuliaClimate/CDSAPI.jl/pull/60#issuecomment-2595223298
and: https://stackoverflow.com/questions/73866908/github-actions-main-repository-secret-not-picked-up-from-pull-request-build

juliohm commented:
> see: [#60 (comment)](https://github.com/JuliaClimate/CDSAPI.jl/pull/60#issuecomment-2595223298) and: https://stackoverflow.com/questions/73866908/github-actions-main-repository-secret-not-picked-up-from-pull-request-build

I always forget. Sorry for the noise. Merging it.
